File: 0859-buddy-strings/0859-buddy-strings.cpp
class Solution {
public:
    bool buddyStrings(string s, string goal) {
        if(s.size() != goal.size()) return false;
        if(s == goal){
            vector<int> freq(26, 0);
            for(auto x : s){
                freq[x - 'a']++;
                if(freq[x-'a'] == 2) return true;
            }
            return false;
        }
        int fIndex = -1, sIndex = -1;
        for(int i=0; i<s.size(); i++){
            if(s[i] != goal[i]){
                if(fIndex == -1) fIndex = i;
                else if(sIndex == -1) sIndex = i;
                else return false;
            }
        }
        if(sIndex == -1) return false;
        return s[fIndex] == goal[sIndex] && s[sIndex] == goal[fIndex];
    }
};

File: 0859-buddy-strings/0859-buddy-strings.py
class Solution:
    def buddyStrings(self, s: str, goal: str) -> bool:
        if len(s) != len(goal): return False
        if s == goal:
            freq = [0] * 26
            for ch in s:
                freq[ord(ch) - ord('a')] += 1
                if freq[ord(ch) - ord('a')] == 2: return True
            return False
        fIndex = -1
        sIndex = -1
        for i in range(len(s)):
            if s[i] != goal[i]:
                if fIndex == -1: fIndex = i
                elif sIndex == -1: sIndex = i
                else: return False
        if sIndex == -1: return False
        return s[fIndex] == goal[sIndex] and s[sIndex] == goal[fIndex]